<p align="center">
<img src="https://github.com/DOH-JDJ0303/vaper-docs/blob/b1478312207e85f963190751ad36a207b4d8cadb/assets/images/vaper-logo_small.png" width="600px" >
</p>

# VAPER: Viral Assembly from Probe-based EnRichment

## Key Features
VAPER is a viral (meta-)assembly pipeline that can:

ðŸ§¬ Assemble genomes from complex samples, supporting multiple assemblies per sample (e.g., co-infections)<br>
ðŸ§¬ Automatically detect and select reference genomes<br>
ðŸ§¬ Predict the taxonomy of each assembly, with an optional viral metagenomic summary<br>
ðŸ§¬ Export reads associated with each assembly for downstream use<br>

While VAPER was originally designed for hybrid capture data (e.g., [Illumina VSP](https://www.illumina.com/products/by-type/sequencing-kits/library-prep-kits/viral-surveillance-panel.html) or [Twist CVRP](https://www.twistbioscience.com/products/ngs/fixed-panels/comprehensive-viral-research-panel)), it has also been used with shotgun metagenomic and tile-amplicon data. It comes stock with comprehensive reference sets for multiple taxa, including influenza A, SARS-CoV-2, and Monkeypox ([full list](https://doh-jdj0303.github.io/vaper-docs/docs/pages/ref_search.html)). Keep on reading to learn more!

## Pipeline Overview
![VAPER Flowchart](https://github.com/DOH-JDJ0303/vaper-docs/blob/b1478312207e85f963190751ad36a207b4d8cadb/docs/media/VAPER-1.0.drawio.png)

## More Information:
See the [docs](https://doh-jdj0303.github.io/vaper-docs/) for more information.

## Acknowledgements
**VAPER would not be possible without the following people:**
- Holly Halstead (WA PHL)
- Pauline Trinh (WA MEP)
- Allison Black (WA MEP)
- Stephanie Lunn (WA MEP)
- Kristen Waterman (WA PHL)
- Brandi Torrevillas (WSU)

VAPER was originally written by Jared Johnson for the Washington State Department of Health.



